### PR TITLE
0.9 RTC8 in-place authorization

### PR DESCRIPTION
**Pull request by trenouf**
I'm doing a PR for this now so @paddy can start code review, but:

1. It includes the commit "RSA8e,RSA9h,RSA10j: do not merge TokenParams and AuthOptions", which also has its own PR. Once that other PR is merged, I can rebase to get rid of that commit.

2. I think this causes a rest test failure, but I haven't worked out what yet as travis is refusing to show me the build log.

### CODE REVIEW

**Comment by trenouf** [comment]
New push fixes the test failure, so this is now ready to go. Although it would be good to get the separate PR for "RSA8e,RSA9h,RSA10j: do not merge TokenParams and AuthOptions" in first.

File: lib/src/main/java/io/ably/lib/http/Http.java [hunk 1]
```diff
@@ -618,6 +635,7 @@ private Response readResponse(HttpURLConnection connection) throws IOException {
 
 		try {
 			response.body = readInputStream(is, response.contentLength);
+			Log.v(TAG, "\n" + new String(response.body));
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/http/Http.java]: Maybe `System.lineSeparator()` if we're being Windows-friendly?

File: lib/src/main/java/io/ably/lib/transport/ConnectionManager.java [hunk 2]
```diff
@@ -94,6 +94,53 @@ public StateIndication(ConnectionState state, ErrorInfo reason, String fallback,
 		}
 	}
 
+	/*************************************
+	 * a class that listens for state change
+	 * events for in-place authorization
+	 *************************************/
+	public static class ConnectionWaiter implements ConnectionStateListener {
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/transport/ConnectionManager.java]: You could avoid the `private Connection` if this wasn't static.

**Comment by trenouf** [comment]
New push fixes both those comments. Ready to go once the travis builds are done.

File: lib/src/main/java/io/ably/lib/http/Http.java [hunk 3]
```diff
@@ -582,6 +593,8 @@ private void writeRequestBody(RequestBody requestBody, HttpURLConnection conn) t
 		conn.setRequestProperty(CONTENT_TYPE, requestBody.getContentType());
 		conn.setRequestProperty(CONTENT_LENGTH, Integer.toString(length));
 		OutputStream os = conn.getOutputStream();
+		if (Log.level <= Log.VERBOSE)
+			Log.v(TAG, "\n" + new String(body));
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/http/Http.java]: You missed this one ....

Comment (trenouf) [inline lib/src/main/java/io/ably/lib/http/Http.java]: Oh bother.

**Comment by trenouf** [comment]
Really fixed this time.

**Comment by paddybyers** [comment]
Thanks

File: lib/src/main/java/io/ably/lib/types/ProtocolMessage.java [hunk 4]
```diff
@@ -205,4 +214,42 @@ public JsonElement serialize(Action action, Type t, JsonSerializationContext ctx
...
+			int fieldCount = unpacker.unpackMapHeader();
+			for(int i = 0; i < fieldCount; i++) {
+				String fieldName = unpacker.unpackString().intern();
+				MessageFormat fieldFormat = unpacker.getNextFormat();
+				if(fieldFormat.equals(MessageFormat.NIL)) { unpacker.unpackNil(); continue; }
+
+				if(fieldName == "accessToken") {
+					accessToken = unpacker.unpackString();
+				} else {
+					System.out.println("Unexpected field: " + fieldName);
```
Comment (mattheworiordan) [inline lib/src/main/java/io/ably/lib/types/ProtocolMessage.java]: Why does this message not go to the logger?

Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/types/ProtocolMessage.java]: That's cut+paste from other existing code that does the same thing. I'm not sure we should even log it tbh - it's legitimate for a message to contain unknown fields when there is a protocol change, for example.

Comment (mattheworiordan) [inline lib/src/main/java/io/ably/lib/types/ProtocolMessage.java]: Yes, but my point is that it should not be going to system out ever anywhere in the library, it should be going to the logger with a suitable log severity.  I am not bothered by which severity you think is appropriate, but I don't like that our library code (not test code) is incorrectly writing to STDOUT. Can we fix anywhere else this happens?

**Comment by paddybyers** [comment]
> Can we fix anywhere else this happens?

We will review all instances and convert to a Logger message or remove.

**Comment by paddybyers** [comment]
> I don't like that our library code (not test code) is incorrectly writing to STDOUT. Can we fix anywhere else this happens?

https://github.com/ably/ably-java/issues/217

**Pull request merged**